crates/cdk/src/wallet/subscription/mod.rs

```diff
@@ -0,0 +1,291 @@
+/// method).
+#[derive(Debug)]
+pub struct SubscriptionClient {
+    new_subscription_notif: mpsc::Sender<SubId>,
+    on_drop_notif: mpsc::Sender<SubId>,
+    subscriptions: Arc<RwLock<HashMap<SubId, WsSubscriptionBody>>>,
+    worker: Option<JoinHandle<()>>,
+}
+/// Active Subscription
+pub struct ActiveSubscription {
+    sub_id: Option<SubId>,
+    on_drop_notif: mpsc::Sender<SubId>,
+    receiver: mpsc::Receiver<NotificationPayload>,
+}
+
+impl ActiveSubscription {
+    fn new(
+        receiver: mpsc::Receiver<NotificationPayload>,
+        sub_id: SubId,
+        on_drop_notif: mpsc::Sender<SubId>,
+    ) -> Self {
+        Self {
+            sub_id: Some(sub_id),
+            on_drop_notif,
+            receiver,
+        }
+    }
+
+    /// Try to receive a notification
+    pub fn try_recv(&mut self) -> Result<Option<NotificationPayload>, Error> {
+        match self.receiver.try_recv() {
+            Ok(payload) => Ok(Some(payload)),
+            Err(mpsc::error::TryRecvError::Empty) => Ok(None),
+            Err(mpsc::error::TryRecvError::Disconnected) => Err(Error::Disconnected),
+        }
+    }
+
+    /// Receive a notification asynchronously
+    pub async fn recv(&mut self) -> Option<NotificationPayload> {
+        self.receiver.recv().await
+    }
+}
+
+impl Drop for ActiveSubscription {
+    fn drop(&mut self) {
+        if let Some(sub_id) = self.sub_id.take() {
+            let _ = self.on_drop_notif.try_send(sub_id);
+        }
+    }
+}
+
+/// Subscription client error
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    /// Url error
+    #[error("Could not join paths: {0}")]
+    Url(#[from] crate::mint_url::Error),
+    /// Disconnected from the notification channel
+    #[error("Disconnected from the notification channel")]
+    Disconnected,
+}
+
+impl SubscriptionClient {
+    /// Create new [`WebSocketClient`]
+    pub fn new(
+        url: MintUrl,
+        http_client: Option<Arc<dyn HttpClientMethods + Send + Sync>>,
+    ) -> Self {
+        let subscriptions = Arc::new(RwLock::new(HashMap::new()));
+        let (new_subscription_notif, new_subscription_recv) = mpsc::channel(100);
+        let (on_drop_notif, on_drop_recv) = mpsc::channel(1000);
+
+        Self {
+            new_subscription_notif,
+            on_drop_notif,
+            subscriptions: subscriptions.clone(),
+            worker: Some(if let Some(http_client) = http_client {
+                Self::http_worker(
+                    http_client,
+                    url,
+                    subscriptions,
+                    new_subscription_recv,
+                    on_drop_recv,
+                )
+            } else {
+                #[cfg(not(all(
+                    not(feature = "http_subscription"),
+                    feature = "mint",
+                    not(target_arch = "wasm32")
+                )))]
+                unreachable!("unreachable code");
```

Comment:
Would it be better to put the mint specific code behind the mint feature instead of using unreachable?

This whole file in in a wallet mod so really there should be no mint relevant code here.